utility/service/service.go

```diff
@@ -122,19 +232,84 @@ func (s servicer) validateRelayChainSupport(relayChain *coreTypes.Identifiable,
 	return nil
 }
 
-// TODO: implement
 // validateApplication makes sure the application has not received more relays than allocated in the current session.
-func (s servicer) validateApplication(meta *coreTypes.RelayMeta, session *coreTypes.Session) error {
-	/*
-		// if maxRelaysPerSession, overServiced := calculateAppSessionTokens(); overServiced {
-			return fmt.Errorf("application %s has exceeded its allocated relays %d for the session %d", meta.ApplicationPublicKey, maxRelaysPerSession)
-		}
-	*/
-	return nil
+func (s *servicer) validateApplication(session *coreTypes.Session, currentHeight int64) error {
+	// IMPROVE: use a function to get current height from the current session
+	servicerAppSessionTokens, err := s.calculateServicerAppSessionTokens(session, currentHeight)
+	if err != nil {
+		return fmt.Errorf("Error calculating servicer tokens for application: %w", err)
+	}
+
+	readCtx, err := s.GetBus().GetPersistenceModule().NewReadContext(currentHeight)
+	if err != nil {
+		return fmt.Errorf("Error getting read context: application %s session number %d: %w", session.Application.PublicKey, session.SessionNumber, err)
+	}
+
+	usedAppSessionTokens, err := readCtx.GetServicerTokenUsage(session)
+	if err != nil {
+		return fmt.Errorf("Error getting servicer token usage: application %s session number %d: %w", session.Application.PublicKey, session.SessionNumber, err)
+	}
+
+	if usedAppSessionTokens == nil || usedAppSessionTokens.Cmp(servicerAppSessionTokens) < 0 {
+		return nil
+	}
+
+	return fmt.Errorf("application %s has exceeded its allocated relays %s for session %d",
+		session.Application.PublicKey,
+		servicerAppSessionTokens,
+		session.SessionNumber)
+}
+
+func (s *servicer) cachedAppTokens(session *coreTypes.Session) *sessionTokens {
+	s.rwlock.RLock()
+	defer s.rwlock.RUnlock()
+
+	return s.totalTokens[session.Application.PublicKey]
+}
+
+// calculateServicerAppSessionTokens return the number of tokens the servicer can use for the application in the current session
+func (s *servicer) calculateServicerAppSessionTokens(session *coreTypes.Session, currentHeight int64) (*big.Int, error) {
```

Comment:
Why not use the `currentHeight` from the session?

Reply:
Thank you for the review. Fixed.

Reply:
Thank you for the review, fixed.